### ISSUE TITLE
Deprecate (remove?) HEADLESS option

### ISSUE THREAD

**Issue by RReverser** [emscripten-core/emscripten#20554]
During #20531 I noticed that we have headless.js & headlessCanvas.js, which led me to discovering HEADLESS option I didn't even know existed:

> If 1, will include shim code that tries to 'fake' a browser environment, in
order to let you run a browser program (say, using SDL) in the shell.
Obviously nothing is rendered, but this can be useful for benchmarking and
debugging if actual rendering is not the issue. Note that the shim code is
very partial - it is hard to fake a whole browser! - so keep your
expectations low for this to work.

It seems it has bare-bones support for emulating browser environment and it was added before browsers provided their own headless variants. Last update (not counting syntax modernisation or licenses) to the relevant code was in 2015.

Nowadays Chrome and Firefox provide headless mode out of the box, which will be much more feature-complete and representative of real-world behaviour for stated goals ("benchmarking and debugging") because, well, it runs in real browser.

Hence, I wonder if it makes sense to deprecate and remove this option and point users to real headless solution instead?

**Comment by sbc100** [comment]
I personally don't use the HEADLESS settting.. but I do vastly prefer working with node compared to `chrome --headless`.   For example just being able to easily see stdout and stderr is really nice.  `chrome --headless` spews all kind of other crap to stdout and IIRC you don't even get to see the console message (or am I wrong about that?)

**Comment by sbc100** [comment]
Having said that I'm always a huge fan of removing unused/underused features!

**Comment by RReverser** [comment]
> I'd be happy to remove this, good idea, but only if it can work as easily as headless does right now. E.g. today

Yeah that sounds sensible.

FWIW another alternative can be to point users to or include [JSDOM](https://github.com/jsdom/jsdom) - which is the most feature-complete project out there doing what HEADLESS option tried to do before (emulating browser environment in Node.js).

That said, I don't like that option as much because there are still differences compared to real browser environment so I don't think any such emulation will help with the stated goals of benchmarking and debugging as much as the real thing.

**Comment by RReverser** [comment]
> For example just being able to easily see stdout and stderr is really nice

Yeah but if you're testing stdout/stderr you can already just use Node.js instead. IIUC, HEADLESS option was more targeted for users who want to automate testing of things like drawing to canvas via SDL and for that browser is the more suitable environment.

**Comment by sbc100** [comment]
Still, given the choice between working with a headless browser and working with node.. I'll always choose node.  So much nicer. 

Having said that, I don't use HEADLESS myself...

**Comment by RReverser** [comment]
> Still, given the choice between working with a headless browser and working with node.. I'll always choose node. So much nicer.

Sure, but the thing is this option is specifically for testing browser-specific APIs which don't exist in Node.

So when you need such testing, alternatives are either provide shims for all possible required browser APIs in Node.js and hope it will behave similarly enough to real browser - which is what HEADLESS option was doing at a very minimum and JSDOM tries to do at maximum, but they both have discrepancies from real browser - or use actual real browser.

IMO when you need a browser, you need a browser. I can totally understand why this option was added when HEADLESS didn't exist, just saying nowadays headless browser is the better environment _for this specific class of tests_. Even Emscripten's own CI uses headless browsers for the same class of tests.

**Comment by kripken** [comment]
Yeah, this is for cases that use browser APIs. I agree browser headless mode is better for it.

Note that another use case back in the day was for benchmarking: we took a browser benchmark like BananaBread and made a headless version that isolated CPU execution (because all GL operations became no-ops), and which was easy to run in the shell. I guess that would still render in browser headless mode, so it would no longer isolate the CPU, but I think that's fine to remove.

**Comment by RReverser** [comment]
I looked into finally removing this, which seems pretty straightforward, but there is some experimental tool in https://github.com/emscripten-core/emscripten/blob/main/tools/experimental/reproduceriter.py that relies on headless.js.

I don't see much docs - does anyone knows how its output is used, whether we can remove the headless dependency from it, and whether anyone still uses it at all?

**Comment by RReverser** [comment]
Hah, looks like it was added all the way back in 2012 by @kripken : https://github.com/emscripten-core/emscripten/commit/ce5523f896495b571e321970be68a8da4b1059fb

What's more interesting, looks like it's not tool that was built to rely on headless.js, it's HEADLESS feature that was extracted out of that tool into its own thing: https://github.com/emscripten-core/emscripten/commit/8263e2377f6cbbb88523aefeb86123e1a6887f1a

Maybe also a candidate for removal?

**Comment by sbc100** [comment]
I say go for it.

**Comment by kripken** [comment]
Yeah, sg to remove that as well.
